interfaces/compatibility/parse.go

```diff
@@ -0,0 +1,253 @@
+	node := &Node{Exp: compatLabel}
+
+	var err error
+	var currDim *CompatDimension
+	var singleInt, leftInt, rightInt uint64
+	tags := map[string]bool{}
+	// Here the lexer makes sure that the sequence of items is correct
+intLoop:
+	for {
+		t := p.peekToken()
+		switch t.Typ {
+		case ItemString:
+			appendDimensionToLabel(compatLabel, currDim)
+
+			if len(t.Val) > 32 {
+				return nil, fmt.Errorf("string is longer than 32 characters: %s", t.Val)
+			}
+			if _, ok := tags[t.Val]; ok {
+				return nil, fmt.Errorf("repeated string in label: %s", t.Val)
+			}
+			currDim = &CompatDimension{Tag: t.Val}
+			tags[t.Val] = true
+		case ItemInteger:
+			singleInt, err = strconv.ParseUint(t.Val, 10, 0)
+			if err != nil {
+				return nil, fmt.Errorf("cannot convert to an integer: %v", err)
+			}
+			currDim.Values = append(currDim.Values,
+				CompatRange{Min: uint(singleInt), Max: uint(singleInt)})
+		case ItemRangeLeftInt:
+			leftInt, err = strconv.ParseUint(t.Val, 10, 0)
+			if err != nil {
+				return nil, fmt.Errorf("cannot convert to an integer: %v", err)
+			}
+		case ItemRangeRightInt:
+			rightInt, err = strconv.ParseUint(t.Val, 10, 0)
+			if err != nil {
+				return nil, fmt.Errorf("cannot convert to an integer: %v", err)
+			}
+			if leftInt > rightInt {
+				return nil, fmt.Errorf("negative range specified: (%d..%d)",
+					leftInt, rightInt)
+			}
+			currDim.Values = append(currDim.Values,
+				CompatRange{Min: uint(leftInt), Max: uint(rightInt)})
+		case ItemEOF, ItemRightParen, ItemAND, ItemOR:
+			break intLoop
+		case ItemLeftParen:
+			return nil, fmt.Errorf("open parenthesis after label")
+		}
+		p.nextToken()
+	}
+	appendDimensionToLabel(compatLabel, currDim)
+	p.labels = append(p.labels, *compatLabel)
+
+	return node, nil
+}
+
+func (p *parser) parseGroupedExpression() (*Node, error) {
+	// Remove '('
+	p.nextToken()
+	p.depth++
+	exp, err := p.parseExpression()
+	if err != nil {
+		return nil, err
+	}
+	// Remove ')'
+	t := p.nextToken()
+	p.depth--
+	if t.Typ != ItemRightParen {
+		return nil, errors.New("expected right parenthesis")
+	}
+
+	return exp, nil
+}
+
+func (p *parser) parseOperator(left *Node) (*Node, error) {
+	op := p.nextToken()
+	node := &Node{Exp: &Operator{Oper: op}, Depth: p.depth, Left: left}
+	t := p.peekToken()
+	switch t.Typ {
+	// Only labels or left parenthesis after operator
+	case ItemString, ItemLeftParen:
+	default:
+		return nil, fmt.Errorf("unexpected token after operator %s: %v", op, t)
+	}
+
+	var err error
+	node.Right, err = p.parseExpression()
+	if err != nil {
+		return nil, err
+	}
+
+	// If our right expression is an operator different to the current one
+	// and it happened at the current depth, return an error as we require
+	// parenthesis to disambiguate in this case.
+	oper, ok := node.Right.Exp.(*Operator)
+	if ok && node.Right.Depth == p.depth &&
+		((op.Typ == ItemAND && oper.Oper.Typ == ItemOR) ||
+			(op.Typ == ItemOR && oper.Oper.Typ == ItemAND)) {
```

Comment:
We already know these are operators so I think we can simplify this to
```suggestion
	if ok && node.Right.Depth == p.depth && op.Typ != oper.Oper.Typ {
```